File: packages/node/src/services/b2c.ts
import { B2cParams } from '../types';
import { HttpClient } from '../http/httpClient';
import { Auth } from '../core/Auth';
import { Helpers } from '../utils/helpers';

export class B2CService {
  private httpClient: HttpClient;
  private auth: Auth;
  private helpers: Helpers;
  private baseUrl: string;

  constructor(httpClient: HttpClient, auth: Auth) {
    this.httpClient = httpClient;
    this.auth = auth;
    this.helpers = new Helpers();
    this.baseUrl = httpClient.getBaseUrl();
  }

  async send(params: B2cParams): Promise<any> {
    const { phonenumber, commandId, amount, remarks, resultUrl, timeoutUrl, shortCodeType = 'B2C' } = params;

    const url = `${this.baseUrl}/mpesa/b2c/v1/paymentrequest`;
    const body = {
      InitiatorName: this.helpers.getConfig('initiatorName'),
      SecurityCredential: this.helpers.generateSecurityCredential(),
      CommandID: commandId,
      Amount: amount,
      PartyA: this.helpers.getConfig('b2cShortcode'),
      PartyB: this.helpers.phoneValidator(phonenumber),
      Remarks: remarks,
      Occassion: '',
      ResultURL: resultUrl || this.helpers.getConfig('callbacks.b2cResultUrl'),
      QueueTimeOutURL: timeoutUrl || this.helpers.getConfig('callbacks.b2cTimeoutUrl'),
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, body, token);
  }

  async validated(params: B2cParams & { idNumber: string }): Promise<any> {
    const { phonenumber, commandId, amount, remarks, idNumber, resultUrl, timeoutUrl, shortCodeType = 'B2C' } = params;

    const url = `${this.baseUrl}/mpesa/b2cvalidate/v2/paymentrequest`;
    const body = {
      InitiatorName: this.helpers.getConfig('initiatorName'),
      SecurityCredential: this.helpers.generateSecurityCredential(),
      CommandID: commandId,
      Amount: amount,
      PartyA: this.helpers.getConfig('b2cShortcode'),
      PartyB: this.helpers.phoneValidator(phonenumber),
      Remarks: remarks,
      Occassion: '',
      OriginatorConversationID: this.helpers.getFormattedTimestamp(),
      IDType: '01',
      IDNumber: idNumber,
      ResultURL: resultUrl || this.helpers.getConfig('callbacks.b2cResultUrl'),
      QueueTimeOutURL: timeoutUrl || this.helpers.getConfig('callbacks.b2cTimeoutUrl'),
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, body, token);
  }
}

File: packages/python/src/mpesa/services/b2c.py
from typing import Optional
from .http.http_client import HttpClient
from .auth import Auth
from .utils import Helpers


class B2CService:
    def __init__(self, http_client: HttpClient, auth: Auth, config):
        self.http_client = http_client
        self.auth = auth
        self.helpers = Helpers(config)
        self.base_url = http_client.get_base_url()

    def send(self, phonenumber: str, command_id: str, amount: int, remarks: str,
             result_url: Optional[str] = None, timeout_url: Optional[str] = None,
             short_code_type: str = 'B2C') -> dict:
        url = f'{self.base_url}/mpesa/b2c/v1/paymentrequest'
        
        body = {
            'InitiatorName': self.helpers.get_config('initiator_name'),
            'SecurityCredential': self.helpers.generate_security_credential(),
            'CommandID': command_id,
            'Amount': amount,
            'PartyA': self.helpers.get_config('b2c_shortcode'),
            'PartyB': self.helpers.phone_validator(phonenumber),
            'Remarks': remarks,
            'Occassion': '',
            'ResultURL': result_url or self.helpers.get_config('callbacks.b2c_result_url'),
            'QueueTimeOutURL': timeout_url or self.helpers.get_config('callbacks.b2c_timeout_url'),
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, body, token)

    def validated(self, phonenumber: str, command_id: str, amount: int, remarks: str,
                  id_number: str, result_url: Optional[str] = None, timeout_url: Optional[str] = None,
                  short_code_type: str = 'B2C') -> dict:
        url = f'{self.base_url}/mpesa/b2cvalidate/v2/paymentrequest'
        
        body = {
            'InitiatorName': self.helpers.get_config('initiator_name'),
            'SecurityCredential': self.helpers.generate_security_credential(),
            'CommandID': command_id,
            'Amount': amount,
            'PartyA': self.helpers.get_config('b2c_shortcode'),
            'PartyB': self.helpers.phone_validator(phonenumber),
            'Remarks': remarks,
            'Occassion': '',
            'OriginatorConversationID': self.helpers.get_formatted_timestamp(),
            'IDType': '01',
            'IDNumber': id_number,
            'ResultURL': result_url or self.helpers.get_config('callbacks.b2c_result_url'),
            'QueueTimeOutURL': timeout_url or self.helpers.get_config('callbacks.b2c_timeout_url'),
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, body, token)
